app/lib/backend/storage_rewrite.rb

```diff
@@ -1,106 +1,155 @@
+      # Execute actual rewriting with Backend, depending on the class.
+      # To be called from the async jobs
+      def rewrite(**kwargs)
+        klass = kwargs[:class_name] || kwargs[:scope]&.klass&.name
+        raise ArgumentError, ':class_name or :scope arguments must be provided' if klass.blank?
 
-      total_count = services.size + cinstances.size + metrics.size + usage_limits.size + bought_cinstances.size
+        rewriter = Backend::StorageRewrite.const_get("#{klass}Rewriter")
+        rewriter.rewrite({ **kwargs, log_progress: log_progress})
+      end
 
-      index = 0
-      progress = -> do
-        percent = ((index + 1) / total_count.to_f) * 100.0
-        yield percent if block_given?
-        index += 1
+      # Schedule all objects for all providers, or execute inline
+      def rewrite_all
+        providers.each do |account|
+          rewrite_provider(account.id)
+        end
+        nil
       end
 
-      services.find_each do |service|
-        service.update_backend_service
-        progress.call
+      # Schedule a single provider or execute inline
+      def rewrite_provider(id)
+        logger.info "#{action} backend storage for provider #{id}..."
 
-        service.service_tokens.find_each(&ServiceTokenService.method(:update_backend))
-      end
+        provider = providers.find_by(id: id)
+        unless provider
+          logger.error("Provider with ID #{id} not found")
+          return
+        end
 
-      metrics.find_each do |metric|
-        metric.sync_backend!
-        progress.call
-      end
+        logger.info "#{action} services for provider #{id}..."
+        process(provider.services)
+
+        logger.info "#{action} buyer applications for provider #{id}..."
+        process(provider.buyer_applications)
+
+        logger.info "#{action} provider applications for provider #{id}..."
+        process(provider.bought_cinstances)
+
+        logger.info "#{action} metrics for provider #{id}..."
+        process(Metric.by_provider(provider))
 
-      usage_limits.find_each do |usage_limit|
-        usage_limit.send(:update_backend_usage_limit)
-        progress.call
+        logger.info "#{action} usage limits for provider #{id}..."
+        process(provider.usage_limits)
       end
 
-      update_cinstance = ->(cinstance) do
-        cinstance.send(:update_backend_application)
-        cinstance.send(:update_backend_user_key_to_application_id_mapping)
-        if cinstance.provider_account
-          cinstance.application_keys.each { |k| k.send(:update_backend_value) }
-          cinstance.referrer_filters.each { |f| f.send(:update_backend_value) }
+      private
+
+      # Enqueue for asynchronous processing in batches (if :enqueuer is provide)
+      # OR execute inline as a whole
+      def process(collection)
+        if enqueuer.respond_to?(:call)
+          collection.in_batches(of: BATCH_SIZE) do |batch|
+            ids = batch.pluck(:id)
+            next if ids.empty?
```

Comment:
I agree here that ids array may never be empty in this loop. Now if you `#compact` it, it might be but this doesn't seem necessary either.

Reply:
Yes, I think you're right, not sure what was my reasoning here :sweat_smile: